File: Source/Game/Management/KittyTime.cs
using System;
using System.Collections.Generic;
using WCSharp.Api;

public class KittyTime
{
    private readonly Action _cachedProgress;
    private Dictionary<int, float> RoundTime { get; set; } = new Dictionary<int, float>();
    private Dictionary<int, float> RoundProgress { get; set; } = new Dictionary<int, float>();
    private AchesTimers ProgressTimer { get; set; } = ObjectPool<AchesTimers>.GetEmptyObject();
    private float TotalTime { get; set; }
    private Kitty Kitty { get; set; }

    public KittyTime(Kitty kitty)
    {
        Kitty = kitty;
        _cachedProgress = () => Progress.CalculateProgress(Kitty);
        Initialize();
    }

    private void Initialize()
    {
        for (int i = 1; i <= Gamemode.NumberOfRounds; i++)
            RoundTime.Add(i, 0.0f);
        for (int i = 1; i <= Gamemode.NumberOfRounds; i++)
            RoundProgress.Add(i, 0.0f);
        PeriodicProgressTimer();
    }

    public void Dispose()
    {
        ProgressTimer.Pause();
        ProgressTimer?.Dispose();
        RoundTime.Clear();
        RoundProgress.Clear();
        RoundTime = null;
        RoundProgress = null;
    }

    private void PeriodicProgressTimer()
    {
        ProgressTimer.Timer.Start(0.2f, true, _cachedProgress);
    }

    #region Time Section

    public float GetRoundTime(int round)
    {
        return RoundTime.ContainsKey(round) ? RoundTime[round] : 0.0f;
    }

    public string GetRoundTimeFormatted(int round)
    {
        return RoundTime.ContainsKey(round) ? Utility.ConvertFloatToTime(RoundTime[round]) : "0:00";
    }

    public float GetTotalTime()
    {
        return TotalTime;
    }

    public string GetTotalTimeFormatted()
    {
        return Utility.ConvertFloatToTime(TotalTime);
    }

    public void SetRoundTime(int round, float time)
    {
        if (RoundTime.ContainsKey(round))
            RoundTime[round] = time;
        SetTotalTime();
    }

    public void IncrementRoundTime(int round)
    {
        if (RoundTime.ContainsKey(round))
            RoundTime[round] += GameTimer.RoundSpeedIncrement;
        SetTotalTime();
    }

    private void SetTotalTime() // Solo Tournament Issue
    {
        TotalTime = 0.0f;
        foreach (var time in RoundTime) // IEnumberable
            TotalTime += time.Value;
    }

    #endregion Time Section

    #region Progress Section

    public float GetRoundProgress(int round)
    {
        return RoundProgress.ContainsKey(round) ? RoundProgress[round] : 0.0f;
    }

    public void SetRoundProgress(int round, float progress)
    {
        RoundProgress[round] = progress;
    }

    public float GetOverallProgress() // Solo Tournament Issue
    {
        float overallProgress = 0.0f;
        foreach (var progress in RoundProgress) // IEnumberable
            overallProgress += progress.Value;
        return overallProgress / RoundProgress.Count;
    }

    #endregion Progress Section
}

File: src/Game/Management/KittyTime.ts
import { Globals } from 'src/Global/Globals'
import { Action } from 'src/Utility/CSUtils'
import { AchesTimers, createAchesTimer } from 'src/Utility/MemoryHandler/AchesTimers'
import { Utility } from 'src/Utility/Utility'
import { Kitty } from '../Entities/Kitty/Kitty'
import { GameTimer } from '../Rounds/GameTimer'
import { Progress } from './Progress'

export class KittyTime {
    private readonly _cachedProgress: Action
    private RoundTime: Map<number, number> = new Map()
    private RoundProgress: Map<number, number> = new Map()
    private ProgressTimer: AchesTimers = createAchesTimer()
    private TotalTime = 0
    private Kitty: Kitty

    public constructor(kitty: Kitty) {
        this.Kitty = kitty
        this._cachedProgress = () => Progress.CalculateProgress(this.Kitty)
        this.Initialize()
    }

    private Initialize = () => {
        for (let i = 1; i <= Globals.NumberOfRounds; i++) this.RoundTime.set(i, 0.0)
        for (let i = 1; i <= Globals.NumberOfRounds; i++) this.RoundProgress.set(i, 0.0)
        this.PeriodicProgressTimer()
    }

    public dispose = () => {
        this.ProgressTimer.pause()
        this.ProgressTimer?.dispose()
        this.RoundTime.clear()
        this.RoundProgress.clear()
    }

    private PeriodicProgressTimer = () => {
        this.ProgressTimer.Timer.start(0.2, true, this._cachedProgress)
    }

    // #region Time Section

    public GetRoundTime(round: number) {
        return this.RoundTime.has(round) ? this.RoundTime.get(round)! : 0.0
    }

    public GetRoundTimeFormatted(round: number) {
        return this.RoundTime.has(round) ? Utility.ConvertFloatToTime(this.RoundTime.get(round)!) : '0:00'
    }

    public GetTotalTime(): number {
        return this.TotalTime
    }

    public GetTotalTimeFormatted(): string {
        return Utility.ConvertFloatToTime(this.TotalTime)
    }

    public SetRoundTime(round: number, time: number) {
        if (this.RoundTime.has(round)) this.RoundTime.set(round, time)
        this.SetTotalTime()
    }

    public IncrementRoundTime(round: number) {
        if (this.RoundTime.has(round))
            this.RoundTime.set(round, this.RoundTime.get(round)! + GameTimer.RoundSpeedIncrement)
        this.SetTotalTime()
    }

    private SetTotalTime = () => {
        // Solo Tournament Issue
        this.TotalTime = 0.0
        for (const [_, time] of this.RoundTime) // IEnumberable
            this.TotalTime += time
    }

    // #endregion Time Section

    // #region Progress Section

    public GetRoundProgress(round: number) {
        return this.RoundProgress.has(round) ? this.RoundProgress.get(round)! : 0.0
    }

    public SetRoundProgress(round: number, progress: number) {
        this.RoundProgress.set(round, progress)
    }

    public GetOverallProgress(): number {
        // Solo Tournament Issue
        let overallProgress = 0.0
        for (const [_, progress] of this.RoundProgress) // IEnumberable
            overallProgress += progress
        return overallProgress / this.RoundProgress.size
    }

    // #endregion Progress Section
}
